Alias `:-webkit-full-screen` pseudo-class to `:fullscreen`

nt1m opened the pull request:
#### c0e3ff49d701c24f709245fc4ed192bb9673a838
<pre>
Alias `:-webkit-full-screen` pseudo-class to `:fullscreen`
<a href="https://bugs.webkit.org/show_bug.cgi?id=266962">https://bugs.webkit.org/show_bug.cgi?id=266962</a>
<a href="https://rdar.apple.com/120335917">rdar://120335917</a>

Reviewed by Darin Adler.

The match functions for :fullscreen and :-webkit-full-screen are compatible with each other.
They differ in some very small edge cases, but for the most common cases they behave identically, alias them to clean up some code.

* LayoutTests/fast/selectors/selector-aliases-expected.txt: Added.
* LayoutTests/fast/selectors/selector-aliases.html: Added.
* LayoutTests/fullscreen/full-screen-css-expected.txt:
* LayoutTests/fullscreen/full-screen-css.html:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesWebkitFullScreenPseudoClass): Deleted.
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Canonical link: <a href="https://commits.webkit.org/272557@main">https://commits.webkit.org/272557@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/de2872841cdb0ee27f4957dfe0d102b305861381

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/32103 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/10822 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/33893 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/34647 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29097 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/32933 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/13179 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8044 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/28674 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/32503 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9131 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/28690 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/7926 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/8096 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/28605 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/35991 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/29191 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29061 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/34207 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8214 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/6165 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32066 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/9855 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/28410 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7497 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/8860 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/8742 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/css/CSSSelector.cpp:
```diff
@@ -530,10 +530,8 @@ String CSSSelector::selectorText(StringView separator, StringView rightSide) con
-            case CSSSelector::PseudoClass::WebKitFullScreen:
-                builder.append(":-webkit-full-screen");
+                // Use serializingValue so alias serializes correctly.
+                builder.append(":"_s, cs->serializingValue());
```
Why do we actually want to roundtrip these? It doesn't matter a whole lot, but I kinda like not having to keep around the non-canonical inputs.

nt1m replied:
I think for the purposes of the Web Inspector and CSSOM, it's a better developer experience if `:-webkit-full-screen` stays `:-webkit-full-screen` instead of unexpectedly changing to `:fullscreen`.

annevk replied:
Is that how CSS aliases generally work? I vaguely recall that they were canonicalized. In DOM we generally canonicalize.

darinadler replied:
Once we decide, I think our decision is best expressed as a test

annevk replied:
https://drafts.csswg.org/selectors/#legacy-aliasing supports my thinking. https://drafts.csswg.org/css-cascade-5/#legacy-name-alias as well, but for properties only.

I agree that a test would be good. Probably can't be a WPT for this particular one as :-webkit-full-screen isn't standardized, but :matches and :-webkit-autofill are.

nt1m replied:
Looks like you're right: https://github.com/w3c/csswg-drafts/issues/5847
I'll add an internal test, `:-webkit-autofill` already has a WPT (which might need to be fixed since it allows serializing to either selectors).

I wonder how well this plays out with WebInspector though, I guess it's not a big deal for CSSOM, but having WebInspector show a selector different than what was authored might throw off a little bit. Perhaps the WebInspector doesn't use CSSOM for this case though.

nt1m replied:
WebInspector doesn't appear to use CSSOM for this case, all good here

nt1m committed:
Alias `:-webkit-full-screen` pseudo-class to `:fullscreen`
https://bugs.webkit.org/show_bug.cgi?id=266962
rdar://120335917

Reviewed by Darin Adler.

The match functions for :fullscreen and :-webkit-full-screen are compatible with each other.
They differ in some very small edge cases, but for the most common cases they behave identically, alias them to clean up some code.

* LayoutTests/fast/selectors/selector-aliases-expected.txt: Added.
* LayoutTests/fast/selectors/selector-aliases.html: Added.
* LayoutTests/fullscreen/full-screen-css-expected.txt:
* LayoutTests/fullscreen/full-screen-css.html:
* Source/WebCore/css/CSSPseudoSelectors.json:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::selectorText const):
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/css/SelectorCheckerTestFunctions.h:
(WebCore::matchesWebkitFullScreenPseudoClass): Deleted.
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Canonical link: https://commits.webkit.org/272557@main
